data/sqldb.go

```diff
@@ -0,0 +1,315 @@
+package data
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/farshidtz/senml/v2"
+	"github.com/linksmart/historical-datastore/common"
+	"github.com/linksmart/historical-datastore/registry"
+	_ "github.com/mattn/go-sqlite3"
+)
+
+// SqlStorage implements a SqlDB storage client for HDS Data API
+type SqlStorage struct {
+	pool *sql.DB
+}
+
+func NewSqlStorage(conf common.DataConf) (storage *SqlStorage, disconnect_func func() error, err error) {
+	storage = new(SqlStorage)
+	storage.pool, err = sql.Open("sqlite3", conf.Backend.DSN) //In future, this shall be made configurable.
+	if err != nil {
+		return nil, nil, err
+	}
+	return storage, storage.Disconnect, err
+}
+
+func btoi(b bool) int {
+	if b {
+		return 1
+	}
+	return 0
+}
+
+func (s *SqlStorage) Submit(data map[string]senml.Pack, sources map[string]*registry.DataStream) error {
+	for ds, pack := range data {
+		valueStrings := make([]string, 0, len(pack))
+		valueArgs := make([]interface{}, 0, len(pack)*2)
+		switch sources[ds].Type {
+		case registry.Float:
+			for _, r := range pack {
+				valueStrings = append(valueStrings, "(?, ?)")
+				valueArgs = append(valueArgs, r.Time, *r.Value)
+			}
+		case registry.String:
+			for _, r := range pack {
+				valueStrings = append(valueStrings, "(?, ?)")
+				valueArgs = append(valueArgs, r.Time, r.StringValue)
+			}
+		case registry.Bool:
+			for _, r := range pack {
+				valueStrings = append(valueStrings, "(?, ?)")
+				valueArgs = append(valueArgs, r.Time, btoi(*r.BoolValue))
+			}
+		case registry.Data:
+			for _, r := range pack {
+				valueStrings = append(valueStrings, "(?, ?)")
+				valueArgs = append(valueArgs, r.Time, r.DataValue)
+			}
+		}
+
+		stmt := fmt.Sprintf("REPLACE INTO [%s] (time, value) VALUES %s",
+			ds, strings.Join(valueStrings, ","))
+		_, err := s.pool.Exec(stmt, valueArgs...)
+		return err
+	}
+
+	return nil
+}
+
+//This function converts a floating point number (which is supported by senml) to a bytearray
+func floatTimeToInt64(senmlTime float64) int64 {
+	return int64(senmlTime * (1e9)) //time.Unix(int64(sec), int64(frac*(1e9))).UnixNano()
+}
+
+//This function converts a bytearray floating point number (which is supported by senml)
+func int64ToFloatTime(timeVal int64) float64 {
+	return float64(timeVal) / 1e9
+}
+
+func ToSenmlTime(t time.Time) float64 {
+	if t.IsZero() {
+		return 0
+	}
+	return int64ToFloatTime(t.UnixNano())
+}
+
+func FromSenmlTime(t float64) time.Time {
+	return time.Unix(0, floatTimeToInt64(t))
+}
+
+//caution: this function modifies the record and baseRecord values
+func denormalizeRecord(record *senml.Record, baseRecord **senml.Record, mask DenormMask) {
```

Comment:
why baseRecord is pointer to pointer?

Reply:
Base record stores the address of first record for future use. Storing just the value does not work unless a deep copy is performed.